Gerrit Batch Changes

varsanojidan opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/50867
Closes https://github.com/sourcegraph/sourcegraph/issues/51929

📺 Loom: https://www.loom.com/share/31d7bdace69d4a5ab8c732339f0a4950

I decided to just close the [Reviews PR](https://github.com/sourcegraph/sourcegraph/pull/52470), and open this up instead because as I was connecting everything I noticed quite a few things that weren't right and I thought it would make more sense to just merge the thing that works directly instead of making it a 2-step process.

This includes basically everything needed for Gerrit Batch Changes, but not webhooks (we don't know if that will be supported yet).  

## Test plan

Added tests

varsanojidan commented on enterprise/internal/batches/reconciler/executor.go:
```diff
@@ -285,6 +292,11 @@ func (e *executor) publishChangeset(ctx context.Context, asDraft bool) (afterDon
 
+	// Gerrit publishes the Change at push time, and therefore
+	if e.ch.ExternalServiceType == extsvc.TypeGerrit {
+		exists = false
```
Wanted to call special attention to this "workaround" here, the Changeset "exists" at this point because it gets created at push time, therefore exists would always return true, and we would always enquere a ChangesetUpdate webhook event instead of the regular publish event.

TBH I'm not sure what ramifications this workaround may cause and wanted to point it out, because as far as I can tell everything works well.

eseliger replied:
could we move this into the changeset layer and return false for the exists flag from CreateChangeset / CreateDraftChangeset instead? That way, this loop would stay free of implementation details. Also, the description in this comment is clearer to me than the code comment so maybe copy that into the code as well? :)

varsanojidan commented on enterprise/internal/batches/reconciler/executor.go:
```diff
@@ -166,6 +168,11 @@ func (e *executor) pushChangesetPatch(ctx context.Context, triggerUpdateWebhook
 
+	// Gerrit relies on a ChangeID that is included in the commit body, so we generate it here.
+	if e.ch.ExternalServiceType == extsvc.TypeGerrit && e.ch.ExternalID == "" {
+		e.ch.ExternalID = gerrit.GenerateRandomChangeID()
```
Also note, that I am generating and setting the Chnage ID here, prior to the push, as it is needed for the commit/push.

eseliger replied:
is this something we could do in the `ChangesetSource`?

varsanojidan replied:
Where did you have in mind? 

The reason I set this here is because this is right before the call to Gitserver to commit/push the changes, so we know that the Changeset hasn't been published yet and therefore we can come up with any Change ID we want, and it gets passed to the CreateChangeset function which just validates that the Change ID we specified actually has a Change attached to it through the Gerrit API (it doesn't create the change, just validates it was created during the push).

eseliger replied:
Heh yeah good question, it's just awkward that we have source specific switches in the generic routine 😬  I've read some more code in this PR, and I've thought the following:
ChangeIDs can really be anything that's random and 41 characters long? If so, could we do something that deterministically generates the change id for the changeset based on the changeset? that way, we could modify the css.GitserverPushOptions to take the changeset and we generate the deterministic ID from it. Then, in CreateChangeset, we can use the same deterministic method.

varsanojidan replied:
I think calculating it deterministically is a great idea!

My latest commit should have addressed all the comments, lmk what you think.

I verified on 2 separate batch changes that the Change IDs are the same at commit time and CreateChangeset time, and were different bw the 2 runs.

varsanojidan replied:
> > I ran through all of the scenarios [#47913 (comment)](https://github.com/sourcegraph/sourcegraph/pull/47913#issuecomment-1439155858), and all of them seem good (with the exception of forks because Gerrit doesn't support it).
> 
> If forks aren't supported, what happens when the flag is set? Does it error, or just ignore the flag?

Just verified that if enforceForks is on, it still successfully makes the Change in Gerrit, might be worth changing the UI to not make it look like its coming from a non-existent fork.
![image](https://github.com/sourcegraph/sourcegraph/assets/20795805/e05eb9bf-4f07-4590-ac22-dd389c24a646)

courier-new replied:
Curious 🤔 I wouldn't have expected that since `GerritSource` isn't a `ForkableChangesetSource` anymore. Does the changeset record in the DB have `external_fork_namespace` set?

varsanojidan replied:
@courier-new sorry, I should have noticed it sooner, the changes I made to the site config got reset before I ran that, and the forks weren't actually enforced.

Re-doing it properly, I see the error in the logs:
[         worker] ERROR worker.batches-reconciler.routines.Handle workerutil/worker.go:381 operation.error {"TraceId": "c3faedac6b84baae4c3bbe69ca3f33a1", "SpanId": "0cc55660f1e507d0", "count": 1, "elapsed": 0.031941959, "error": "forking is enabled, but the changeset source does not support forks"}

which makes sense.

courier-new replied:
Oh got it, cool! Does that log get surfaced in the UI at all, then? I assume the changeset fails to publish? I think it's fine for it to fail (it wouldn't really be _enforcing_ forks if it just pushed the change to the target repo anyway, eh? 😛), but it'd be great if it was clear from the UI why it failed.

varsanojidan replied:
This is the error in the UI
![image](https://github.com/sourcegraph/sourcegraph/assets/20795805/584d8bf0-2b4c-4f2f-b780-ee108e36ad76)

courier-new replied:
Oh sweet that's perfect then! Yay!

varsanojidan commented:
I ran through all of the scenarios [here](https://github.com/sourcegraph/sourcegraph/pull/47913#issuecomment-1439155858), and all of them seem good with the exception of forks because Gerrit doesn't support it.

eseliger reviewed (commented):
> I ran through all of the scenarios https://github.com/sourcegraph/sourcegraph/pull/47913#issuecomment-1439155858, and all of them seem good (with the exception of forks because Gerrit doesn't support it).

If forks aren't supported, what happens when the flag is set? Does it error, or just ignore the flag?

varsanojidan commented:
> could we move this into the changeset layer and return false for the exists flag from CreateChangeset / CreateDraftChangeset instead? That way, this loop would stay free of implementation details. Also, the description in this comment is clearer to me than the code comment so maybe copy that into the code as well? :)

I think my brain must have lagged and not actually finished the comment lol, yeah I'll update it. I can give it a shot for sure, the reason I was hesitant was because the Change technically exists at the time you call CreateChangeset, so I wasn't sure if it might mess with some other logic, but I'll try it out and validate that everything behaves as expected.

sourcegraph-buildkite commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | -0.00% (-0.35 kb) | -0.00% (-0.35 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/iv/gitserver-support-gerrit-bc/compare-report.html;tab=live_object) for a full comparison between the commits 63b3cc3 and aa7cc2e or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

courier-new commented on enterprise/internal/batches/types/changeset.go:
```diff
@@ -588,9 +589,9 @@ func (c *Changeset) Body() (string, error) {
 	case *gerritbatches.AnnotatedChange:
 		// Gerrit doesn't really differentiate between title/description.
-		return m.Subject, nil
+		return m.Change.Subject, nil
```
I noticed for `Title` you `strings.Cut` to just the first line, if there's multiple. Would it make sense here to do something similar? If it's a one-liner, just keep `m.Change.Subject`, but if there's multiple, take only lines 2+?

varsanojidan replied:
Gerrit is kinds weird where there isn't really a title, just the commit message, so in terms of our usage, I think it makes sense for the title to be the first line of the commit message, and then the Changeset Body to be the full commit message, what do you think?

courier-new replied:
Sure, that probably makes sense too!

courier-new commented on enterprise/internal/batches/types/changeset_event.go:
```diff
@@ -108,14 +108,20 @@ const (
 	ChangesetEventKindAzureDevOpsPullRequestApproved                ChangesetEventKind = "azuredevops:pullrequest:approved"
 	ChangesetEventKindAzureDevOpsPullRequestApprovedWithSuggestions ChangesetEventKind = "azuredevops:pullrequest:approved_with_suggestions"
-	ChangesetEventKindAzureDevOpsPullRequesReviewed                 ChangesetEventKind = "azuredevops:pullrequest:reviewed"
+	ChangesetEventKindAzureDevOpsPullRequestReviewed                ChangesetEventKind = "azuredevops:pullrequest:reviewed"
```
Oops good catch 👀

courier-new commented on internal/extsvc/gerrit/testdata/golden/GetChangeReviews.json:
```diff
@@ -0,0 +1,19 @@
+    "Code-Review": "+1"
+   },
+   "_account_id": 1000034,
+   "name": "Kelli Rockwell",
+   "email": "kelli@sourcegraph.com"
```
Wowowow I've been memorialized in a Gerrit VCR test, I can die happy now! 🎉

courier-new reviewed (approved):
👍👍👍👍
Thank you for adding all the tests here too!!!

varsanojidan commented on client/web/src/enterprise/batches/settings/AddCredentialModal.tsx:
```diff
@@ -59,14 +59,6 @@ const scopeRequirements: Record<ExternalServiceKind, JSX.Element> = {
             permissions.
         </span>
     ),
-    [ExternalServiceKind.GERRIT]: (
-        // TODO: @varsanojidan change this after spike
-        <span>
-            with <Code>Organization:All accessible organizations</Code>, <Code>Code:Full</Code>,{' '}
-            <Code>Code:Status</Code>, <Code>Pull Request Threads:Read & Write</Code>, and <Code>User Profile:Read</Code>{' '}
-            permissions.
-        </span>
-    ),
```
Nope 😅, I added it back as returning just an empty span.

varsanojidan commented:
@courier-new thank you for the amazing feedback, I believe I should've addressed all of your suggestions in my most recent PR

courier-new commented on cmd/gitserver/server/patch.go:
```diff
@@ -217,11 +218,12 @@ func (s *Server) createCommitFromPatch(ctx context.Context, req protocol.CreateC
-		cmd = exec.CommandContext(ctx, "git", "commit", "-m", fmt.Sprintf("%q", message), "-m", fmt.Sprintf("Change-Id: %s", req.Gerrit.ChangeID))
+	formattedMessages := []string{"commit"}
+	for _, m := range req.CommitInfo.Messages {
+		formattedMessages = append(formattedMessages, "-m", m)
```
Ought we still quote each part of the commit message, as before? Otherwise I think certain characters like dashes in the commit message can get misinterpreted by the shell as additional command arguments and cause errors:

```suggestion
		formattedMessages = append(formattedMessages, "-m", fmt.Sprintf("%q", m))
```

varsanojidan replied:
It seems like this change causes it not to parse the Change ID correctly because of the quotes, so I think maybe we can move this formatting to the BuildCommitOpts and then just not do it for the Change ID message in Gerrit's Build commit opts? what do you think?

courier-new replied:
Huh interesting! Can you tell why it can't parse the Change ID part of the commit message exactly? I'm okay doing that if we really have to, it just feels a bit bad putting that responsibility on the consumer.

courier-new commented on enterprise/internal/batches/sources/gerrit.go:
```diff
@@ -211,6 +212,14 @@ func (s GerritSource) MergeChangeset(ctx context.Context, cs *Changeset, _ bool)
+	opts.PushRef = &pushRef
+	opts.CommitInfo.Messages = append(opts.CommitInfo.Messages, "Change-Id: "+GenerateGerritChangeID(*changeset))
+	return opts
+}
```
:chefs-kiss:

courier-new commented on cmd/gitserver/server/patch.go:
```diff
@@ -47,6 +47,7 @@ func (s *Server) handleCreateCommitFromPatchBinary(w http.ResponseWriter, r *htt
 }
 
 func (s *Server) createCommitFromPatch(ctx context.Context, req protocol.CreateCommitFromPatchRequest) (int, protocol.CreateCommitFromPatchResponse) {
+	fmt.Printf("Req: %+v\n", req)
```
```suggestion
```

courier-new commented on cmd/gitserver/server/patch.go:
```diff
@@ -217,7 +218,12 @@ func (s *Server) createCommitFromPatch(ctx context.Context, req protocol.CreateC
 	}
 
-	cmd = exec.CommandContext(ctx, "git", "commit", "-m", fmt.Sprintf("%q", message))
+	formattedMessages := []string{"commit"}
```
Haha nittiest nit but for a second I thought you dropped the `commit` part of the command because I didn't realize you initialized the `formattedMessages` slice with it. Might it be clearer for this slice to start empty, and then you `cmd = exec.CommandContext(ctx, "git", "commit", formattedMessages...)` instead?

varsanojidan replied:
Unfortunately, Go doesn't let you specify it as multiple args, I changed it to this

`cmd = exec.CommandContext(ctx, "git", append([]string{"commit"}, formattedMessages...)...)`

Which I don't know if it is much cleaner than before, what do you think?

courier-new replied:
Ah I see, ew. 😬 Yeah I guess those are both pretty ugly. How about we leave it the way you had it, then, but instead of calling it `formattedMessages` we call it `gitArgs` or something?

varsanojidan replied:
Sounds like a plan 😅

courier-new commented on cmd/gitserver/server/patch.go:
```diff
@@ -276,7 +282,12 @@ func (s *Server) createCommitFromPatch(ctx context.Context, req protocol.CreateC
 
 	if req.Push != nil {
-		cmd = exec.CommandContext(ctx, "git", "push", "--force", remoteURL.String(), fmt.Sprintf("%s:%s", cmtHash, ref))
+		pushRef := ref
+		// If the push is to a Gerrit project,we need to push to a magic ref.
+		if req.PushRef != nil && *req.PushRef != "" {
+			pushRef = *req.PushRef
+		}
```
Nit: Could we move this, the unique-ref-finding logic going on up at 138, and line 70 where `ref` is originally defined all closer together, so it's clearer how the value of `ref` gets assigned and reassigned depending on `req` options? That might make it clearer that `req.PushRef` takes precedence over `req.TargetRef` and that `req.UniqueRef` only matters if `req.PushRef` isn't provided?

varsanojidan replied:
I moved it in my recent PR, lmk if this what you had in mind 😄

courier-new replied:
I like it, but I think this might make lines 138-158 (the stuff with `if req.UniqueRef` and `ref = ensureRefPrefix(ref)`) ineffectual. I assume you assign the new variable `pushRef` to `ref` because you didn't want to interfere with how `ref` is used anywhere else in this method, but I actually scanned the rest of it, and `ref` is literally _only_ used in the context of the `git push` command. 😂 Makes your life a bit simpler. But right now since you've swapped that `git push` arg to `pushRef`, and the older pre-existing logic in 138-158 only applied to `ref`, that code path is just a dead-end. It's a little weird that all that ref-assigning logic was sooo spread out in this method beforehand, ahaha, so I think this is a good opportunity to clean that up a bit. Does that all make sense?

courier-new commented on enterprise/internal/batches/sources/common.go:
```diff
@@ -96,6 +97,8 @@ type ChangesetSource interface {
 	// merge. If the changeset cannot be merged, because it is in an unmergeable
 	// state, ChangesetNotMergeableError must be returned.
 	MergeChangeset(ctx context.Context, ch *Changeset, squash bool) error
+	// BuildCommitOpts builds the options needed to push the change to the code host based on the specific ChangesetSource.
```
Maybe 

```suggestion
	// BuildCommitOpts builds the CreateCommitFromPatchRequest needed to commit and push the change to the code host.
```

courier-new reviewed (commented):
Great changes, left a couple more minor comments.

Another question, since you didn't go this far in your Loom recording: What is the behavior for updating a changeset, like when you go back and execute an updated batch spec that changes the `title` or produces a different diff and you need to commit and push the new changes? My impression is that it would create a brand new Change on Gerrit, since `GenerateGerritChangeID` would produce a different ID now. Do we update the tracking okay, i.e. is there still only one Change associated with the batch change after that update? Do we do anything with the old Change on Gerrit/should we?

varsanojidan commented on enterprise/internal/batches/sources/gerrit.go:
```diff
@@ -89,30 +94,67 @@ func (s GerritSource) LoadChangeset(ctx context.Context, cs *Changeset) error {
+		return false, errors.Wrap(err, "getting change")
+	}
+
+	pr, err := s.client.GetChange(ctx, changeID)
```
yep, it returns an empty object 🤷

varsanojidan commented:
@courier-new I believe I should have addressed all of the comments, the last thing I need to do is validate the scenario when updating the Changeset now that we use the deterministic Change ID approach. I'll update here tomorrow when I look into it.

varsanojidan commented:
@courier-new now that we do the deterministic ChangeID, updating the Batch Change (title and script), as you predicted, does create a new Change on the Gerrit code host, but the tracking is not updated locally, so the Batch Change still has the old Change.

As you called out, the creation of the new Change is happening because of the deterministic ChangeID generation. My question is, what behavior do we want here? Are we ok with it just creating a new change and we update the Changeset to point to the new Change? Do we want it to still be attached to the old change in which case the deterministic Change ID might be an issue here?

varsanojidan commented:
@courier-new I got it to work almost as expected [here](https://www.loom.com/share/0aa48c5e437d4c7b9905dfe8a040c253), it just required messing around a bit with Gerrit's BuildCommitOpts func to use an existing ChangeID instead of generating one if ti exists 😄.
